### PR TITLE
#405 reuse http client

### PR DESCRIPTION
**Pull request by jonnybee**
Changed SendGridClient to keep/reuse HttpClient instance 
Removed obsolete properties and parameters (that cannot be re-set on the HttpClient)
Modified unit tests to pass custom headers to SendGridClient constructor
Change in Examples to avoid exception when there is no groups and correct text message.

### CODE REVIEW

File: src/SendGrid/SendGridClient.cs [hunk 1]
```diff
@@ -129,168 +144,25 @@ public enum Method
...
         /// </summary>
-        /// <param name="client">Client object ready for communication with API</param>
         /// <param name="request">The parameters for the API call</param>
         /// <param name="cancellationToken">Cancel the asynchronous call</param>
         /// <returns>Response object</returns>
-        public async Task<Response> MakeRequest(HttpClient client, HttpRequestMessage request, CancellationToken cancellationToken = default(CancellationToken))
+        public async Task<Response> MakeRequest(HttpRequestMessage request, CancellationToken cancellationToken = default(CancellationToken))
         {
-            HttpResponseMessage response = await client.SendAsync(request, cancellationToken).ConfigureAwait(false);
+            HttpResponseMessage response = await client.SendAsync(request, cancellationToken);
```
Comment (thinkingserious) [inline src/SendGrid/SendGridClient.cs]: Why drop the `.ConfigureAwait(false);`? I believe we had async issues before without this call. Please see this thread: https://github.com/sendgrid/sendgrid-csharp/issues/362

**Comment by jonnybee** [comment]
Done. Learned a new thing  now about why use ConfigureAwait(false) in libraries.

File: src/SendGrid/SendGridClient.cs [hunk 2]
```diff
@@ -61,12 +34,69 @@ public class SendGridClient
         /// <param name="host">Base url (e.g. https://api.sendgrid.com)</param>
         /// <param name="requestHeaders">A dictionary of request headers</param>
         /// <param name="version">API version, override AddVersion to customize</param>
-        /// <param name="urlPath">Path to endpoint (e.g. /path/to/endpoint)</param>
         /// <returns>Interface to the SendGrid REST API</returns>
-        public SendGridClient(IWebProxy webProxy, string apiKey, string host = "https://api.sendgrid.com", Dictionary<string, string> requestHeaders = null, string version = "v3", string urlPath = null)
-            : this(apiKey, host, requestHeaders, version, urlPath)
+        public SendGridClient(IWebProxy webProxy, string apiKey, string host = null, Dictionary<string, string> requestHeaders = null, string version = "v3")
```
Comment (thinkingserious) [inline src/SendGrid/SendGridClient.cs]: Please re-add the urlPath parameter so we can avoid a breaking change. It allows for the user to specify a default endpoint path.

File: src/SendGrid/SendGridClient.cs [hunk 3]
```diff
@@ -76,25 +106,10 @@ public SendGridClient(IWebProxy webProxy, string apiKey, string host = "https://
         /// <param name="host">Base url (e.g. https://api.sendgrid.com)</param>
         /// <param name="requestHeaders">A dictionary of request headers</param>
         /// <param name="version">API version, override AddVersion to customize</param>
-        /// <param name="urlPath">Path to endpoint (e.g. /path/to/endpoint)</param>
         /// <returns>Interface to the SendGrid REST API</returns>
-        public SendGridClient(string apiKey, string host = null, Dictionary<string, string> requestHeaders = null, string version = "v3", string urlPath = null)
+        public SendGridClient(string apiKey, string host = null, Dictionary<string, string> requestHeaders = null, string version = "v3")
```
Comment (thinkingserious) [inline src/SendGrid/SendGridClient.cs]: Please re-add the urlPath parameter so we can avoid a breaking change. It allows for the user to specify a default endpoint path.

File: src/SendGrid/SendGridClient.cs [hunk 4]
```diff
@@ -23,35 +22,9 @@ namespace SendGrid
...
-        /// <summary>
-        /// The request media type.
-        /// </summary>
-        private string mediaType;
-
-        /// <summary>
-        /// The web proxy.
-        /// </summary>
-        private IWebProxy webProxy;
+        private readonly string version;
```
Comment (thinkingserious) [inline src/SendGrid/SendGridClient.cs]: Please include summary comments.

**Comment by jonnybee** [comment]
Done.

File: src/SendGrid/SendGridClient.cs [hunk 5]
```diff
@@ -63,10 +52,67 @@ public class SendGridClient
         /// <param name="version">API version, override AddVersion to customize</param>
         /// <param name="urlPath">Path to endpoint (e.g. /path/to/endpoint)</param>
         /// <returns>Interface to the SendGrid REST API</returns>
-        public SendGridClient(IWebProxy webProxy, string apiKey, string host = "https://api.sendgrid.com", Dictionary<string, string> requestHeaders = null, string version = "v3", string urlPath = null)
-            : this(apiKey, host, requestHeaders, version, urlPath)
```
Comment (thinkingserious) [inline src/SendGrid/SendGridClient.cs]: Not sure why this line was removed.

File: src/SendGrid/SendGridClient.cs [hunk 6]
```diff
@@ -79,22 +125,8 @@ public SendGridClient(IWebProxy webProxy, string apiKey, string host = "https://
         /// <param name="urlPath">Path to endpoint (e.g. /path/to/endpoint)</param>
         /// <returns>Interface to the SendGrid REST API</returns>
         public SendGridClient(string apiKey, string host = null, Dictionary<string, string> requestHeaders = null, string version = "v3", string urlPath = null)
+            : this(null, apiKey, host, requestHeaders, version, urlPath)
```
Comment (thinkingserious) [inline src/SendGrid/SendGridClient.cs]: .. and then this line added here.

**Comment by jonnybee** [comment]
New version create the HttpClient in ctor and this is done in the ctor that accepts all parameters (WebProxy). 

Previous version of this ctor would call the other ctor (without the webproxy parameter) and then just set the WebPoxy property.

**Comment by thinkingserious** [comment]
That makes sense, and thanks for responding so quickly. My main concern is that this may cause a breaking change because the function signatures are no longer the same. I will double check the tests and make sure code written against the previous versions of those constructors do not break.

**Comment by thinkingserious** [comment]
Thanks again for the awesome PR! Please take a moment to fill out [this form]( https://docs.google.com/forms/d/e/1FAIpQLSelP4uF-wGJebxVYq1DYw5WzHomEUauab3YiqrshyCkRs0fqw/viewform#start=invite) so that we can send you some swag :)

**Pull request merged**